cluster-autoscaler/debuggingsnapshot/debugging_snapshotter.go

```diff
@@ -0,0 +1,222 @@
+	ResponseHandler(http.ResponseWriter, *http.Request)
+	// IsDataCollectionAllowed checks the internal State of the snapshotter
+	// to find if data can be collected. This can be used before preprocessing
+	// for the snapshot
+	IsDataCollectionAllowed() bool
+	// Flush triggers the flushing of the snapshot
+	Flush()
+	// Cleanup clears the internal data beans of the snapshot, readying for next request
+	Cleanup()
+}
+
+// NewDebuggingSnapshotter returns a new instance of DebuggingSnapshotter
+func NewDebuggingSnapshotter(isDebuggerEnabled bool) DebuggingSnapshotter {
+	state := SNAPSHOTTER_DISABLED
+	if isDebuggerEnabled {
+		klog.Infof("Debugging Snapshot is enabled")
+		state = LISTENING
+	}
+	return &DebuggingSnapshotterImpl{
+		State:             &state,
+		Mutex:             &sync.Mutex{},
+		DebuggingSnapshot: &DebuggingSnapshotImpl{},
+		FlushTrigger:      make(chan bool),
+	}
+}
+
+// ResponseHandler is the impl for request handler
+func (d *DebuggingSnapshotterImpl) ResponseHandler(w http.ResponseWriter, r *http.Request) {
+	// If the debugging snapshot is disabled return 404.
+	// This is a preventive measure in case the handler is listening without the flag
+	// Although in the main we don't listen if the flag is not set
+	if *d.State == SNAPSHOTTER_DISABLED {
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	// checks if the handler is in the correct State to accept a new snapshot request
+	if *d.State != LISTENING {
+		klog.Errorf("Debugging Snapshot is currently being processed. Another snapshot can't be processed")
+		w.WriteHeader(http.StatusTooManyRequests)
+		w.Write([]byte("Another debugging snapshot request is being processed. Concurrent requests not supported"))
+		return
+	}
+
+	d.Mutex.Lock()
+	// set the State to trigger enabled, to allow workflow to collect data
+	*d.State = TRIGGER_ENABLED
+	d.Mutex.Unlock()
+
+	// waiting until an exit State is set. Until then it is considered that CA is processing
+	// the snapshot request
+	<-d.FlushTrigger
+
+	d.DebuggingSnapshot.SetTimestamp(time.Now().In(time.UTC))
+	body, isError := d.DebuggingSnapshot.GetAll()
+	if isError {
+		w.WriteHeader(http.StatusInternalServerError)
+	} else {
+		w.WriteHeader(http.StatusOK)
+	}
+	w.Write(*body)
+
+	d.Mutex.Lock()
+	// reset the debugging State to receive a new snapshot request
+	*d.State = LISTENING
+	d.Mutex.Unlock()
+}
+
+// IsDataCollectionAllowed encapsulate the check to know if data collection is currently active
+// The should be used by setters and by any function that is contingent on data collection State
+// before doing extra processing.
+// e.g. If you want to pre-process a particular State in cloud-provider for snapshot
+// you should check this func in the loop before doing that extra processing
+func (d *DebuggingSnapshotterImpl) IsDataCollectionAllowed() bool {
+	statesAllowed := map[DebuggingSnapshotterState]bool{
+		DATA_COLLECTED:        true,
+		START_DATA_COLLECTION: true,
+	}
+
+	return statesAllowed[*d.State]
```

Comment:
nit: Having a map just to compare one value with two constants seems to be an overkill.

Reply:
You don't need if-else, just `return *d.State == START_DATA_COLLECTION || *d.State == DATA_COLLECTED`. Up to you though.